### PR TITLE
Implement hook to subscribe getblocks blockindex latestblock

### LINKED ISSUE #924: Implement hook to subscribe get blocks / block index / latest block
It should perform GraphQL subscribe to get info from block table back.

### PR DESCRIPTION
**Pull request by prin-r**
Fix #924  #926

### CODE REVIEW

File: scan/src/utils/HexUtils.re [hunk 1]
```diff
@@ -1,10 +1,9 @@
...
-  let len = hexstr->String.length;
-  let prefix = len >= 2 ? hexstr->String.sub(0, 2) : "";
-
-  switch (prefix) {
-  | "0x"
-  | "0X" => hexstr->String.lowercase_ascii->String.sub(2, len - 2)
-  | _ => hexstr->String.lowercase_ascii
-  };
-};
+let normalizeHexString = hexstr =>
```
Comment (sorawit) [inline scan/src/utils/HexUtils.re]: Why do we need to change this?

File: scan/src/subscriptions/Sub.re [hunk 2]
```diff
@@ -14,6 +14,11 @@ let default = (result, value) =>
   | _ => value
   };
 
+let toOption =
```
Comment (sorawit) [inline scan/src/subscriptions/Sub.re]: Let's not have this function defined here. I don't think we want it long term.

File: scan/src/subscriptions/BlockSub.re [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  result
+  |> Sub.map(_, x => x##blocks_aggregate##aggregate |> Belt_Option.getExn |> (y => y##count));
+};
+
+let getProposer = (block: t, validators: list(ValidatorHook.Validator.t)) =>
+  validators
+  ->Belt_List.keep(validator => validator.consensusPubkey |> PubKey.toAddress == block.proposer)
+  ->Belt_List.get(0);
+
+let getProposerMoniker = (block: t, validators: list(ValidatorHook.Validator.t)) =>
```
Comment (sorawit) [inline scan/src/subscriptions/BlockSub.re]: Why do we need this? I think it's far better to have proposer as foreign key to validator table and get moniker from there.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/utils/HexUtils.re (hunk 1 above)
Comment (prin-r) [reply scan/src/utils/HexUtils.re]: Just make it work for prefix 0x,0X and now we have \\x

Comment (sorawit) [reply scan/src/utils/HexUtils.re]: Please add a good comment explaining what this function does. It's nontrivial for someone reading this now to understand the intention of this function.

Also add more tests to the test for `normalizeHexString`.

File: scan/src/utils/ID.re [hunk 4]
```diff
@@ -61,6 +61,10 @@ module IDCreator = (RawID: RawIDSig) => {
     fun
     | ID(id) => RawID.prefix ++ string_of_int(id);
 
+  let toInt =
```
Comment (sorawit) [inline scan/src/utils/ID.re]: Why add this? Can't you just use `toString`? If the comma is important, let's add `toStringPretty` or something, though I don't think we want that. Exposing `int` out is just not safe and antipattern.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/subscriptions/BlockSub.re [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    },
+};
+
+module MultiConfig = [%graphql
+  {|
+  subscription Blocks($limit: Int!, $offset: Int!) {
+    blocks(limit: $limit, offset: $offset, order_by: {height: desc}) @bsRecord {
+      height @bsDecoder(fn: "GraphQLParser.int64")
+      hash: block_hash @bsDecoder(fn: "GraphQLParser.hash")
+      proposer @bsDecoder(fn: "Address.fromHex")
```
Comment (sorawit) [inline scan/src/subscriptions/BlockSub.re]: Don't think this is the right thing to do. @taobun can you confirm?

File: scan/src/utils/HexUtils.re (hunk 1 above)
Comment (sorawit) [reply scan/src/utils/HexUtils.re]: Also if you are changing this, then we can also change `GraphQLParser.hash` to not do the `substr`, but I feel like we should just not do this change.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/subscriptions/BlockSub.re (hunk 5 above)
Comment (taobun) [reply scan/src/subscriptions/BlockSub.re]: In database we store consensus validator address, so if you want validator address and moniker just query more data
```
validator{
      consensus_address
      operator_address
      moniker
    }
```

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/__tests__/JsBuffer_test.re [hunk 6]
```diff
@@ -13,6 +13,16 @@ describe("Expect JsBuffer to work correctly", () => {
     |> toEqual(fromHex("f23391b5dbf982e37fb7dadea64aae21cae4c172"))
   );
 
+  test("should be able to convert upper case hex with 0X to lower case without 0X", () =>
+    expect(fromHex("0XF23391B5DBF982E37FB7DADEA64AAE21CAE4C172"))
+    |> toEqual(fromHex("f23391b5dbf982e37fb7dadea64aae21cae4c172"))
+  );
+
+  test({j|should be able to convert upper case hex with \\\\x to lower case without \\\\x|j}, () =>
```
Comment (sorawit) [inline scan/__tests__/JsBuffer_test.re]: Why are you testing \\\\x? I think the data from database only has \x?

File: scan/src/utils/HexUtils.re [hunk 7]
```diff
@@ -1,10 +1,13 @@
...
-  let len = hexstr->String.length;
-  let prefix = len >= 2 ? hexstr->String.sub(0, 2) : "";
-
-  switch (prefix) {
-  | "0x"
-  | "0X" => hexstr->String.lowercase_ascii->String.sub(2, len - 2)
-  | _ => hexstr->String.lowercase_ascii
-  };
-};
+// This function receive any string and fliter the latest part
```
Comment (sorawit) [inline scan/src/utils/HexUtils.re]: Make sure the sentence is grammatically correct and contains no spelling errors. At least you should pass it through grammarly.

**Review by sorawit — DISMISSED** [review]

File: scan/src/subscriptions/BlockSub.re [hunk 8]
```diff
@@ -0,0 +1,118 @@
+open ValidatorSub.Mini;
+
+type aggregate_t = {count: int};
+
+type transactions_aggregate_t = {aggregate: option(aggregate_t)};
+
+type internal_t = {
+  height: int,
```
Comment (taobun) [inline scan/src/subscriptions/BlockSub.re]: Can we use ID.Block.t?

File: scan/src/subscriptions/BlockSub.re [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  hash: Hash.t,
+  validator: ValidatorSub.Mini.t,
+  timestamp: MomentRe.Moment.t,
+  transactions_aggregate: transactions_aggregate_t,
+};
+
+type t = {
+  height: int,
```
Comment (taobun) [inline scan/src/subscriptions/BlockSub.re]: ditto

File: scan/src/pages/BlockIndexPage.re [hunk 10]
```diff
@@ -109,7 +103,7 @@ let make = (~height: int) => {
     <Row>
       <Col size=1.8>
         {switch (blockOpt) {
-         | Some(block) => <InfoHL info={InfoHL.Count(block.numTxs)} header="TRANSACTIONS" />
+         | Some(block) => <InfoHL info={InfoHL.Count(block.txn)} header="TRANSACTIONS" />
```
Comment (taobun) [inline scan/src/pages/BlockIndexPage.re]: I thought we have many ways to handle None/NoData. Now I think we use let%Sub in top of make function and return Sub.resolve. Look example at DataSourcePage.re

File: scan/src/pages/BlockHomePage.re [hunk 11]
```diff
@@ -68,102 +69,86 @@ let renderBody = ((block, proposer): (BlockHook.Block.t, option(ValidatorHook.Va
...
+              spacing={Text.Em(0.06)}
+              height={Text.Px(15)}
+              nowrap=true
+              block=true
+              color=Colors.gray7
+            />
+            <div className=Styles.seperatedLine />
+            <Text
+              value={
+                switch (blocks->Belt_Array.get(0)) {
```
Comment (taobun) [inline scan/src/pages/BlockHomePage.re]: Change to blockCounts

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/__tests__/HexUtils_test.re [hunk 12]
```diff
@@ -21,11 +21,19 @@ describe("Expect Hash to work correctly", () => {
     |> toBe("aaa22e077492cbad414098ebd98aa8dc1c7ae8d9")
   );
 
+  test({j|should be able to strip \\x out and return the lower case of substring after \\x|j}, () =>
```
Comment (evilpeach) [inline scan/__tests__/HexUtils_test.re]: Should it substring after \x?

File: scan/__tests__/JsBuffer_test.re [hunk 13]
```diff
@@ -13,6 +13,16 @@ describe("Expect JsBuffer to work correctly", () => {
     |> toEqual(fromHex("f23391b5dbf982e37fb7dadea64aae21cae4c172"))
   );
 
+  test("should be able to convert upper case hex with 0X to lower case without 0X", () =>
+    expect(fromHex("0XF23391B5DBF982E37FB7DADEA64AAE21CAE4C172"))
+    |> toEqual(fromHex("f23391b5dbf982e37fb7dadea64aae21cae4c172"))
+  );
+
+  test({j|should be able to convert upper case hex with \\x to lower case without \\x|j}, () =>
```
Comment (evilpeach) [inline scan/__tests__/JsBuffer_test.re]: It should be \x.

File: scan/src/pages/BlockIndexPage.re (hunk 10 above)
Comment (prin-r) [reply scan/src/pages/BlockIndexPage.re]: Waiting for TxSub

File: scan/src/contexts/GlobalContext.re [hunk 14]
```diff
@@ -1,7 +1,7 @@
 type t = {
   financial: PriceHook.Price.t,
-  latestBlock: BlockHook.Block.t,
-  latestBlocks: list(BlockHook.Block.t),
+  latestBlock: BlockSub.t,
```
Comment (taobun) [inline scan/src/contexts/GlobalContext.re]: Remove latestBlock and latestBlocks from global context.

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/subscriptions/ValidatorSub.re [hunk 15]
```diff
@@ -0,0 +1,7 @@
+module Mini = {
+  type t = {
+    consensusAddress: Address.t,
```
Comment (sorawit) [inline scan/src/subscriptions/ValidatorSub.re]: Let's leave this as string.

File: scan/src/utils/HexUtils.re [hunk 16]
```diff
@@ -1,10 +1,14 @@
...
 let normalizeHexString = hexstr => {
-  let len = hexstr->String.length;
-  let prefix = len >= 2 ? hexstr->String.sub(0, 2) : "";
-
-  switch (prefix) {
-  | "0x"
-  | "0X" => hexstr->String.lowercase_ascii->String.sub(2, len - 2)
-  | _ => hexstr->String.lowercase_ascii
-  };
+  hexstr
```
Comment (sorawit) [inline scan/src/utils/HexUtils.re]: Good effort but let's not change it this time.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/subscriptions/TxSub.re [hunk 17]
```diff
@@ -416,6 +456,18 @@ module TxCountConfig = [%graphql
 |}
 ];
 
+module TxCountBySenderConfig = [%graphql
+  {|
+  subscription Transaction($sender: String!) {
+    transactions_aggregate(where: {sender: {_eq: $sender}}) {
+      aggregate{
```
Comment (sorawit) [inline scan/src/subscriptions/TxSub.re]: nit: code format

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]
Looks good!

**Pull request merged**